[RUM-11630] Support SVG in Session Replay

cdn34dd opened the pull request:
<div align="center">
  <img src="https://github.com/user-attachments/assets/7f7fe5ed-bc7a-4b5e-8fbb-9ca682bbcdd9" width="50%" />
</div>


### What does this PR do?

  This PR adds support for tracking SVG images in React Native Session Replay.

  This PR is composed of 3 main components that work together to create a workflow that enables SVGs to be captured in Session Replay:

  - New Babel Plugin Feature (`@datadog/mobile-react-native-babel-plugin` package)
  - New Metro Plugin (`@datadog/mobile-react-native-session-replay` package)
  - New CLI Utility (`@datadog/mobile-react-native-babel-plugin` package)

## Babel Plugin

  The Babel plugin feature is the core of this PR, containing most of the complexity. Its main responsibilities are to:

  - Extract SVG nodes from the user's codebase during the build process
  - Convert SVG nodes into a web-compliant format, which involves:
    - Updating tag and property names (e.g., strokeWidth → stroke-width)
    - Removing React Native-specific metadata
    - Converting styling information from React Native style objects into proper web SVG properties
    - Extracting transform properties from style objects
  - Save extracted SVGs as optimized assets on the file system using SVGO optimization
  - Properly identify SVG code nodes so they can be tracked by Session Replay on the native side

### What kind of SVGs does the plugin handle?

  To correctly identify and extract SVGs in a user's codebase, the plugin differentiates between three types of SVGs and handles each appropriately:

  **1. Inline SVGs**

  SVGs built directly in the codebase using elements from the react-native-svg library (e.g., `<Svg>`, `<Path>`, `<Circle>`). These require the most complex logic for full extraction and transformation.

  Example:
```tsx
<Svg width="100" height="100">
   <Circle cx="50" cy="50" r="45" fill="blue" />
</Svg>
```

  **2. External SVGs (Local)**

  SVGs loaded from the user's filesystem. These are typically already in web-compliant format, but the plugin needs to properly identify them in the user's code since they can be named anything.

  Example:
```tsx
import MyIllustration from './assets/illustration.svg';

<MyIllustration width="100" height="100" />
```

  **3. External SVGs (Remote)**

  SVGs loaded from external servers using the SvgUri component from react-native-svg.

  Example:
```tsx
 <SvgUri uri="https://example.com/icon.svg" />
```

  ### What kind of SVGs do we still need to support?

  The plugin currently handles direct usage of `react-native-svg` elements and loading SVG files directly. The main area for future expansion is adding support for popular SVG libraries built on top of `react-native-svg`, such as:

  - react-native-feather
  - iconoir-react-native
  - @fluentui/react-native-icons

  The groundwork to support these libraries is already in place, as the logic should be similar to handling local SVGs (treating library-generated SVGs as specialized local SVG cases).

  ### Areas of Improvement

  Some areas for future enhancement that didn't make it into this first release:

  - Improved tracking of local SVGs when using aliased paths (e.g., @assets/icon.svg)
  - Enhanced variable tracking for dynamically referenced SVG sources
  - Extended style-to-SVG property conversion for edge cases where users set style properties directly on SVG elements
  - Support for SVG data URLs. While the plugin itself supports it, the native Session Replay SDKs on Android and iOS still need to add support for it. 
  - Better error handling and reporting during the transformation process

  All of these improvements are feasible with sufficient time allocation.

  ## Metro Plugin

  The Metro plugin (`withSessionReplayAssetBundler`) integrates with React Native's Metro bundler to automatically aggregate SVG assets during development and production builds.

  ### Key responsibilities:

  - Monitors the assets directory for new SVG files created by the Babel plugin during development
  - Merges individual SVG files into a single optimized binary format (`assets.bin`) with an accompanying JSON index (`assets.json`) for efficient native lookup
  - Automatically triggers asset aggregation on Metro events such as `bundle_build_done` and `transformer_load_done`
  - Uses file watching with debouncing to efficiently handle rapid asset changes during hot reload

### How it works:

  The Metro plugin wraps the Metro config's reporter and watches for new .svg files in the
  packages/react-native-session-replay/assets/ directory. When new SVG assets are detected (either from the Babel plugin or the CLI utility), it packs them into:
  - `assets.bin` - A binary file containing all SVG data concatenated together
  - `assets.json` - An index mapping SVG IDs to their byte offset and length in the binary file

  This approach allows the native SDKs to efficiently load SVG assets by reading specific byte ranges from a single file rather than managing hundreds of individual files.

  ## CLI Utility

  The CLI utility (`npx datadog-generate-sr-assets`) provides a way to pre-generate all Session Replay SVG assets before native builds.

  Key responsibilities:

  - Scans the entire codebase for React components containing SVG elements
  - Processes files through the Babel plugin with SVG tracking enabled
  - Extracts and saves SVG assets to the Session Replay package's assets directory
  - Generates the final assets.bin and assets.json files for native platform consumption

### When to use it:

  The CLI utility should be run after installing dependencies and before building iOS/Android apps:

  **After installing dependencies**
  `yarn install`

  **Generate SVG assets**
  `npx datadog-generate-sr-assets`

  **Then build native apps**
  `cd ios && pod install && cd ..`

  This ensures that native asset references are available during the build process, particularly important for iOS where the
  assets need to be included in the Xcode project.

  ### What it does:

  1. Clears any existing assets to ensure a fresh state
  2. Scans all `.js`, `.jsx`, `.ts`, and `.tsx` files (excluding `node_modules`, test files, etc.)
  3. Transforms each file using the Babel plugin to extract SVG assets
  4. Merges all extracted SVGs into the binary format
  5. Reports the number of assets generated

## How does this all work together?

  Here's the complete workflow for integrating SVG support in your React Native app:

  **1. Setup the Babel Plugin**

  Configure the Babel plugin in your `babel.config.js` to enable SVG tracking:
```js
  module.exports = {
    presets: ['module:@react-native/babel-preset'],
    plugins: [
      [
        '@datadog/mobile-react-native-babel-plugin',
        {
          sessionReplay: {
             // This controls the SVG asset extraction activation
             // if set to false (default), no assets are extracted
            svgTracking: true
          }
        }
      ]
    ]
  };
```

 **2. Setup the Metro Plugin**

  Configure the Metro plugin in your `metro.config.js` to enable asset bundling:

```js
const { withSessionReplayAssetBundler } = require('@datadog/mobile-react-native-session-replay/metro');

module.exports = withSessionReplayAssetBundler({ /* your existing Metro config */ });
```

  **3. Generate Assets Before Native Builds**

  Run the CLI utility after installing dependencies and before building native apps:

  **Install dependencies**
  `yarn install`

  **Generate Session Replay SVG assets**
  `npx datadog-generate-sr-assets`

  **Build iOS (if needed)**
  `cd ios && pod install && cd ..`

  **Now you can run your app**
  `yarn ios`
  **or**
  `yarn android`

 **4. Development Workflow**

  During development, the Metro plugin automatically handles new SVG assets created by the Babel plugin:

  - Write your components with SVG elements from `react-native-svg`
  - Metro bundler detects the new assets via the Babel plugin transformation
  - The Metro plugin watches for new .svg files and automatically merges them
  - Assets are immediately available for Session Replay recording

  This ensures that SVGs are seamlessly captured in Session Replay without requiring manual asset management.

cdn34dd committed:
Add constants' values definining SVG's elements and properties to target

cdn34dd committed:
Add required SVG utility functions to convert RN properties to web compliant properties

cdn34dd committed:
Add various babel node processing function specific to SVG processing

cdn34dd committed:
Add fs functions to ensure the assets directory is correctly set up

cdn34dd committed:
Add babel plugin support for inline SVG conversion to web compliant format

cdn34dd committed:
Add SVGO optimization step

cdn34dd committed:
Add Support for Local SVGs

- Refactor code to allow multiple types of SVG elements

cdn34dd committed:
Add babel plugin Support for SVGUri

cdn34dd committed:
Add script to generate assets folder in session replay package

cdn34dd committed:
Add metro plugin to session replay package

cdn34dd committed:
Ensure session replay package is found before running babel SVG processing

cdn34dd committed:
Update 'DdPrivacyView' spec and attributes (JS/Native)

cdn34dd committed:
Set android's session replay assets directory

cdn34dd committed:
Refactor logic for identifying SVG views natively

- Wrap identified SVG nodes with a `SessionReplayView.Privacy` view, so
we can pass extra information to the native side (e.g., `width`,
`height`, `hash`, ....)

cdn34dd committed:
Add function to babel plugin that saves extracted SVG assets to disk

cdn34dd committed:
Add babel plugin option to handle session replay functionality

- Add `sessionReplay` option to the plugin, allowing SVG tracking to be
enabled or disabled.

cdn34dd committed:
Add new session replay react-native android mapper

- Add a new session replay mapper targeting SVG views

cdn34dd committed:
Add new session replay react-native iOS node recorder

- Add new session replay node recorder targeting SVG views

cdn34dd committed:
Add cli script to babel plugin to handle initial resource generation

sbarrio commented on packages/react-native-babel-plugin/README.md:
```diff
@@ -40,10 +40,19 @@ You can configure the plugin to adjust how it processes your code, giving you co
 | Option               | Type   | Default | Description |
 |-----------------------|--------|---------|-------------|
 | `actionNameAttribute` | string | –       | The chosen attribute name to use for action names. |
+| `sessionReplay`          | object | –       | Session Replay options configuration. |
```
Nit: Perhaps we should leave it as `"Session Replay configuration."`

sbarrio commented on packages/react-native-babel-plugin/package.json:
```diff
@@ -39,8 +42,14 @@
+        "@babel/types": "^7.27.1",
+        "fast-glob": "^3.3.3",
+        "svgo": "^4.0.0",
+        "uuid": "^11.1.0"
```
Nit: Is there any specific reason not to use 13.0.0?

cdn34dd replied:
Probably not, I mostilkey just copied the dependency from another package.json we have in our repo 😁

sbarrio commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+                    ],
+                    '@babel/preset-react'
+                ],
+                // Don't generate actual output, we just the asset generation
```
`we just the asset generation`

Typo?

sbarrio commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+            offset = fs.statSync(binPath).size;
+        } catch (err) {
+            index = {};
+            offset = 0;
```
Should'n we recreate these files if the reading process fails?, also, shouldn't we handle each one separately?, we won't run this code if either the bin of json files don't exist, is that correct?

marco-saia-datadog replied:
@sbarrio I think it might not be necessary, as the binary file is written using `fs.appendFileSync`, which [creates the file if it does not exist](https://nodejs.org/api/fs.html#fsappendfilesyncpath-data-options), and the JSON file is always overwritten with a new one, with `fs.writeFileSync`

cdn34dd replied:
Yes that is the goal.

sbarrio replied:
That makes more sense indeed. Thanks!

sbarrio commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/constants.ts:
```diff
@@ -0,0 +1,292 @@
+];
+
+export const rnSvgArrayAttributeValues = [
+    'stroke-dasharray', // this attribute is already in the web compliant format, sice the RN to web conversion already occured when this is used
```
Typo: `sice` => `since`

sbarrio commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/constants.ts:
```diff
@@ -0,0 +1,292 @@
+    'startOffset',
+    'stdDeviation',
+    'stitchTiles',
+    // 'strokeDasharray', //
```
Should we remove or uncomment this?

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+            }
+            if (fs.existsSync(jsonPath)) {
+                fs.unlinkSync(jsonPath);
+            }
```
The checks seem redundant, as you have already checked the existence of the files above. To refactor this, while still keeping it safe, I would unlink both files, but also make sure to catch exceptions for these operation as well. To make it cleaner you could wrap this in a private utility function.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+        }
+        if (fs.existsSync(jsonPath)) {
+            fs.unlinkSync(jsonPath);
+        }
```
If you use a private utility function, like suggested above, you can re-use it here.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+    for (const f of files) {
+        if (f === binName || f === jsonName) {
+            continue;
+        }
```
This logic assumes that every file in `assets/` (except the .bin and .json) should be processed, but that might not always be true. If extra files accidentally end up in the directory (like .DS_Store, .gitkeep, or future additions), they'll be picked up and processed incorrectly, potentially breaking the build (although the current logic is crash safe, but users would still see a bunch of `[mergeSvgAssets]` errors).

It would be safer to explicitly filter only the file types we expect (`.svg`), for example:

```ts
const files = fs.readdirSync(assetsDir)
  .filter(f => f.endsWith('.svg'))
  .sort();
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    if (errorCount > 0) {
+        console.log(`${errorCount} files had errors`);
```
I’d use `console.warn` here instead.

It might also be helpful to support a `--verbose` flag for this command, so we can log additional details about the processed files when needed, that would make it easier to troubleshoot reported issues.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+    mergeSvgAssets(assetsPath);
+
+    if (errorCount > 0) {
+        console.log('Some files had errors but asset generation completed.');
```
**ultra-nit**: the message could be slightly improved:

`console.log('Asset generation completed with some errors.');`

or

`console.log('Asset generation finished, but some files encountered errors.');`

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+        console.log('Some files had errors but asset generation completed.');
+    }
+
+    console.log('Your assets are now ready to be used by session replay.');
```
**ultra-nit**:

```suggestion
    console.log('Your assets are now ready to be used by Session Replay.');
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/cli/generate-sr-assets.ts:
```diff
@@ -0,0 +1,206 @@
+ *   or
+ *   npx datadog-generate-sr-assets
+ */
+function generateSRAssets() {
```
**nit:** I'd suggest using the full name `generateSessionReplayAssets()` here for consistency with our naming conventions.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/constants.ts:
```diff
@@ -0,0 +1,292 @@
+
+export const xmlNamespace = ['xmlns', 'http://www.w3.org/2000/svg'];
+
+export const svgElements = [
```
**important**: It does not seem like this is used anywhere.

cdn34dd replied:
Nice catch, we don't indeed, but we probably should confirm that the converted name is correct 😅

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/constants.ts:
```diff
@@ -0,0 +1,292 @@
+    'preserveAspectRatio',
+    'primitiveUnits',
+    'refX',
+    'refY',
```
I would refactor these attribute arrays into Sets to get O(1) lookup time, since they're only used with `.includes(item)`.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/handlers/UriSvgHandler.ts:
```diff
@@ -0,0 +1,88 @@
+
+import type { SvgHandler } from './SvgHandler';
+
+export class UriSvgHandler implements SvgHandler {
```
**nit**: JS Doc for the class
```ts
/**
 * Handles extraction and transformation of SVG data from JSX elements that reference SVGs by URI.
 * 
 * The `UriSvgHandler` inspects a JSXElement, extracts its `uri` attribute to locate the SVG source,
 * and captures its dimensional attributes (e.g., `width`, `height`) for further processing.
 * 
**/
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/handlers/UriSvgHandler.ts:
```diff
@@ -0,0 +1,88 @@
+                if (t.isStringLiteral(attr.value)) {
+                    uri = attr.value.value;
+                }
+                console.log('attr: ', JSON.stringify(attr, null, 2));
```
**required**: remove `console.log` to reduce noise. It might be worth to keep it by improving its phrasing if you decide to implement some kind of `verbose` option for troubleshooting.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/handlers/RNSvgHandler.ts:
```diff
@@ -0,0 +1,260 @@
+
+import type { SvgHandler } from './SvgHandler';
+
+export class RNSvgHandler implements SvgHandler {
```
*nit*: JS doc for the class

```ts
/**
 * Internal handler that transforms React Native–style SVG JSXElements into
 * web-compatible SVG output during Babel processing.
 *
 * The `RNSvgHandler` normalizes tag names and attributes, extracts
 * width/height dimensions, consolidates transform attributes, and ensures
 * correct SVG namespace declarations.
 */
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/index.ts:
```diff
@@ -0,0 +1,381 @@
+    length: number;
+};
+
+export class ReactNativeSVG {
```
**nit**: JS doc for the class

```ts
/**
 * Internal processor responsible for detecting, transforming, and wrapping
 * React Native SVG components for use with Session Replay.
 *
 * This class scans the project for `.svg` imports, builds a mapping between
 * JSX identifiers and SVG files, and transforms JSX SVG nodes into
 * optimized, web-compatible SVG markup. Each transformed element is then
 * wrapped in a `SessionReplayView.Privacy` component with metadata used by
 * the native Session Replay layer.
 */
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/index.ts:
```diff
@@ -0,0 +1,381 @@
+     * and config files.
+     */
+    buildSvgMap() {
+        // IMPROVEMENT: Support aliased paths
```
**suggestion**: we should follow the convention of writing TODOs in the form:

```ts
// TODO: Short description (JIRA-TICKET-ID)
```
This ensures we don't lose track of TODOs.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/index.ts:
```diff
@@ -0,0 +1,381 @@
+        });
+
+        if (!alreadyImported) {
+            const importDecl = t.importDeclaration(
```
**question**: Is this injection logic compatible with all module types (CommonJS, ESM)?
Will it insert a `require()` statement instead of an import when running on CommonJS modules, or does it always inject an import declaration regardless of the module system?

cdn34dd replied:
It always injects a import declaration, but given the scope is only react native, this is ok, because everything goes through metro and it handles the mixing ESM and commonJS (if present) correctly.

The only scenario I can think of that could be an issue is if someone would try to run the plugin on code that isn't handled by metro, then maybe they could have issues. But that's not a use case for use.

If at some point though we wanted to make this module-aware we would need some logic to check the import/require statements and infer which module system to target or something else along these lines.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/attributes.ts:
```diff
@@ -0,0 +1,278 @@
+    if (dimensionAttributes.includes(attrName)) {
+        const result = getJSXAttributeData(t, attr);
+
+        // IMPROVEMENT: If it's a variable try to find it in the closure
```
**suggestion**: as mentioned in one of my previous comments 

```ts
// TODO: If it's a variable try to find it in the closure (JIRA-TICKET-ID)
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/attributes.ts:
```diff
@@ -0,0 +1,278 @@
+ * @param transformsArray - Accumulator array for transform operations.
+ * @returns True if the transform list was extracted successfully.
+ */
+export function handleJoinedTranformAttributes(
```
**typo**:

```suggestion
export function handleJoinedTransformAttributes(
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/attributes.ts:
```diff
@@ -0,0 +1,278 @@
+            return true;
+        }
+    }
+    return false;
```
Can be simplified by using early returns, removing unnecessary nesting:

```ts
export function handleJoinedTranformAttributes(
    t: typeof Babel.types,
    attr: Babel.types.JSXAttribute,
    attrName: string,
    transformsArray: { name: string; value: string | number }[] = []
) {
    if (attrName !== 'transform' ||
        !t.isJSXExpressionContainer(attr.value) ||
        !t.isArrayExpression(attr.value.expression)
    ) {
        return false;
    }

    const transformList = evaluateStaticNode(t, attr.value.expression);

    if (!Array.isArray(transformList)) {
        return false;
    }

    for (const entry of transformList) {
        if (!entry || typeof entry !== 'object' || Array.isArray(entry)) {
            continue; // <-- personal preference, nesting with an if is perfectly fine if you prefer it
        }

        for (const key of Object.keys(entry)) {
            transformsArray.push({
                name: key,
                value: entry[key]
            });
        }
    }

    return true;
}
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/fs.ts:
```diff
@@ -0,0 +1,150 @@
+    try {
+        fs.accessSync(assetsPath, fs.constants.F_OK);
+    } catch (error) {
+        fs.mkdirSync(assetsPath);
```
**suggestion**:

```suggestion
        try {
          fs.mkdirSync(assetsPath);
        } catch (err) {
          console.error(`Failed to create directory "${assetsPath}": ${err.message}`);
          throw err;
        }
```
`mkdirSync` can still throw, and we should handle it

cdn34dd replied:
This function is not used, can just be removed.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/svg.ts:
```diff
@@ -0,0 +1,211 @@
+    const value = convertArrayExpressionToArray(t, expression, []);
+
+    if (Array.isArray(value)) {
+        switch (attrName) {
+            case 'gradientTransform':
+            case 'matrix':
+                return `matrix(${value.join(' ')})`;
+
+            case 'values':
+                return value.join(';');
+
+            case 'points': {
+                const data = value
+                    .map(v => (Array.isArray(v) ? v.join(',') : v))
+                    .join(' ');
+
+                return data;
+            }
+
+            default:
+                return value.join(' ');
+        }
+    }
+
+    return null;
+}
```
**nit**: personal preference, as it removes nesting of the switch-case

```ts
export function convertAttributeArrayValue(
    t: typeof Babel.types,
    expression: Babel.types.ArrayExpression,
    attrName: string
) {
    const value = convertArrayExpressionToArray(t, expression, []);
    if (!Array.isArray(value)) {
        return null;
    }

    switch (attrName) {
        case 'gradientTransform':
        case 'matrix':
            return `matrix(${value.join(' ')})`;

        case 'values':
            return value.join(';');

        case 'points': {
            return value
                .map(v => (Array.isArray(v) ? v.join(',') : v))
                .join(' ');
        }

        default:
            return value.join(' ');
    }
}
```

Also, I would add the return type in the function signature as `string | null`

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/processing/svg.ts:
```diff
@@ -0,0 +1,211 @@
+        }
+
+        // Ignore unsupported elements (identifiers, spreads)
+        console.warn('Unsupported array element in SVG prop:', element);
```
**nit**: to be consistent with the other logs

```suggestion
        console.warn('[convertArrayExpressionToArray] Unsupported array element in SVG prop:', element);
```

marco-saia-datadog commented on packages/react-native-babel-plugin/src/libraries/react-native-svg/utils.ts:
```diff
@@ -0,0 +1,365 @@
+
+    // Shadows / elevation → box-shadow
+    if (
+        style.shadowColor ||
+        style.shadowOffset ||
+        style.shadowOpacity != null ||
+        style.shadowRadius != null
+    ) {
+        const color = rgba(
+            style.shadowColor || '#000',
+            style.shadowOpacity ?? 1
+        );
+        const off: RNShadow = style.shadowOffset || {};
+        const blur = style.shadowRadius != null ? style.shadowRadius : 0;
+
+        css['box-shadow'] = `${addPx(off.width || 0)} ${addPx(
+            off.height || 0
+        )} ${addPx(blur)} 0 ${color}`;
+    } else if (style.elevation != null) {
+        const e = style.elevation;
+        css['box-shadow'] = `0 ${addPx(e)} ${addPx(1.5 * e)} 0 rgba(0,0,0,0.3)`;
+    }
+
+    // Pointer events
+    if (style.pointerEvents === 'none') {
+        css['pointer-events'] = 'none';
+    } else if (style.pointerEvents === 'auto') {
+        css['pointer-events'] = 'auto';
+    }
+
+    // Other
+    if (style.direction) {
+        css.direction = style.direction;
+    }
+    if (style.aspectRatio != null) {
+        css['aspect-ratio'] = String(style.aspectRatio);
+    }
+
+    return css;
+}
```
**suggestion**: direct mappings could be grouped, as this is a 278 lines function 😨

```ts
const directMappings = {
  position: 'position',
  display: 'display',
  overflow: 'overflow',
  opacity: 'opacity',
  backgroundColor: 'background-color',
  borderColor: 'border-color',
  borderStyle: 'border-style',
  borderTopColor: 'border-top-color',
  borderRightColor: 'border-right-color',
  borderBottomColor: 'border-bottom-color',
  borderLeftColor: 'border-left-color',
  flex: 'flex',
  flexDirection: 'flex-direction',
  flexWrap: 'flex-wrap',
  justifyContent: 'justify-content',
  alignItems: 'align-items',
  alignContent: 'align-content',
  alignSelf: 'align-self',
  direction: 'direction',
};
```

And applied with:

```ts
for (const [key, cssKey] of Object.entries(directMappings)) {
  if (style[key] != null) css[cssKey] = style[key];
}
```

And the same goes for the dimensional props:

```ts
const dimensionalProps = [
  'width', 'height', 'minWidth', 'minHeight',
  'maxWidth', 'maxHeight', 'top', 'bottom', 'left', 'right',
  'margin', 'marginTop', 'marginRight', 'marginBottom', 'marginLeft',
  'padding', 'paddingTop', 'paddingRight', 'paddingBottom', 'paddingLeft',
  'borderWidth', 'borderRadius', 'borderTopLeftRadius', 'borderTopRightRadius',
  'borderBottomRightRadius', 'borderBottomLeftRadius',
  'borderTopWidth', 'borderRightWidth', 'borderBottomWidth', 'borderLeftWidth',
  'flexBasis', 'gap', 'rowGap', 'columnGap',
];
for (const k of dimensionalProps) {
  if (style[k] != null) css[kebabCase(k)] = addPx(style[k]);
}
```

Further optimizations could surely be performed for the other keys as well, as I see repeating patterns.

cdn34dd replied:
While I agree this function is quite large and could be optimized, I don't think this should be done now, as the work here is more than just organizing things by certain patterns. 

The main optimization here should be to better adapt these styling conversions to the Session Replay Player to make sure the final render is as close as possible between app and replay.

Only after we know that should we think about organizing the code better in logical groups. But yes, something like what you suggest here would make sense.

marco-saia-datadog commented on packages/react-native-babel-plugin/src/utils/nodeProcessing.ts:
```diff
@@ -288,3 +288,253 @@ export function toExpression(
+        }
+
+        // Any other type at top-level is not a valid style object
+        return null;
+    }
+
+    return null;
```
**nit**: seems redundant to have `return null` inside the last if

marco-saia-datadog commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactNativeInternalCallback.kt:
```diff
@@ -17,7 +21,134 @@ import com.facebook.react.bridge.ReactContext
+     * their offset and length inside the `assets.bin` binary file.
+     *
+     * @return A [JSONObject] representing the parsed asset index, or `null` if
+     * the file could not be read or contained invalid JSON.
+     */
+    private fun loadAssetsJson(): JSONObject? {
+        return try {
+            val jsonText = reactContext.assets.open(ASSETS_JSON)
+                .bufferedReader()
+                .use { it.readText() }
+            JSONObject(jsonText)
+        } catch (e: IOException) {
+            null
+        } catch (e: JSONException) {
+            null
+        }
+    }
```
**suggestion**: I think we should log a warning if we expect the assets JSON file to be there. Would it be possible to have a flag to determine whether `svgTracking` is enabled or not?

cdn34dd replied:
We do not, this is a feature available only through the babel plugin, it's not expected for the file to not be preset.

marco-saia-datadog commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactNativeInternalCallback.kt:
```diff
@@ -17,7 +21,134 @@ import com.facebook.react.bridge.ReactContext
+        val offset = entry.optInt("offset", -1)
+        val length = entry.optInt("length", -1)
+
+        if (offset < 0 || length <= 0) {
+            return null
+        }
+
+        return try {
+            readAssetBytes(reactContext, offset, length)
+        } catch (e: Exception) {
+            null
+        }
```
**suggestion**: as per my previous comment, I would improve error handling if possible

marco-saia-datadog commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,162 @@
+                let bundle = Bundle(for: SvgViewRecorder.self)
+                if let url = bundle.url(forResource: "assets", withExtension: "bin") {
+                    do {
+                        let data = try Data(contentsOf: url)
```
**blocking**: as discussed, we should use an `InputStream` or `FileHandle` to avoid loading the whole binary file in memory, and take advantage of the binary + JSON approach 

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/index.ts:
```diff
@@ -0,0 +1,75 @@
+
+let watching = false;
+
+export function withSessionReplayAggregator(metroConfig: any): any {
```
**note**: I am not a big fan of this name. I would go for something different, like:

- `withSessionReplayAssetBundler`
- `withSessionReplaySvgMerger`
- `withSvgAssetBundler` - more generic but it would still be imported from our SR package

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/index.ts:
```diff
@@ -0,0 +1,75 @@
+    const assetsDir = path.resolve(__dirname, '../../../assets');
+
+    if (!fs.existsSync(assetsDir)) {
+        fs.mkdirSync(assetsDir, { recursive: true });
+    }
+
+    const debounceMerge = debounce(() => {
+        try {
+            mergeSvgAssets(assetsDir);
+        } catch (error) {
+            console.warn('[SessionReplayAggregator] merge failed:', error);
+        }
+    }, 300);
+
+    // Prevent potential multiple watchers if metro loads the config multiple times
+    if (!watching) {
+        watching = true;
+        chokidar
+            .watch(assetsDir, {
+                // Skip events for files that already exist
+                ignoreInitial: true,
+                depth: 0,
+                awaitWriteFinish: {
+                    // Wait 200ms after last change
+                    stabilityThreshold: 200,
+                    // Check every 50ms
+                    pollInterval: 50
+                }
+            })
+            .on('add', file => {
+                if (!file.endsWith('.svg')) {
+                    return;
+                }
+
+                debounceMerge();
+            });
+    }
```
**nit**: I'd extract timings into constants, something like:

- `MERGE_DEBOUNCE_MS`
- `WATCH_STABILITY_THRESHOLD_MS`
- `WATCH_POLL_INTERVAL_MS`

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/processing.ts:
```diff
@@ -0,0 +1,99 @@
+type SvgIndex = Record<string, SvgIndexEntry>;
+
+export function mergeSvgAssets(assetsDir: string) {
+    // const assetsDir = 'node_modules/@datadog/mobile-react-native-session-replay/assets'
```
**nit**: Remove comment from here, and use JS doc on the function itself to explain what path it will look for

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/processing.ts:
```diff
@@ -0,0 +1,99 @@
+            }
+            if (fs.existsSync(jsonPath)) {
+                fs.unlinkSync(jsonPath);
+            }
```
Refer to my comment on `generate-sr-assets.ts`:

>The checks seem redundant, as you have already checked the existence of the files above. To refactor this, while still keeping it safe, I would unlink both files, but also make sure to catch exceptions for these operation as well. To make it cleaner you could wrap this in a private utility function.

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/processing.ts:
```diff
@@ -0,0 +1,99 @@
+        if (fs.existsSync(jsonPath)) {
+            fs.unlinkSync(jsonPath);
+        }
+    }
```
**blocking**:

All FS functions should be wrapped in try and catch, especially `unllinkSync`.

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/processing.ts:
```diff
@@ -0,0 +1,99 @@
+            }`
+        );
+    }
+}
```
All comments for `generate-sr-assets.ts` apply here as well. 

Would it be possible to share the logic between the two files?

cdn34dd replied:
Yeah it's kinda redundant, but they live in different packages, so there's probably no easy way that's 100% safe to share the logic. Also I changed the logic a bit for the CLI, so that it always generates fresh asset files, so that anything that was not used anymore got cleaned up.

marco-saia-datadog reviewed (changes requested):
Amazing work 💯  the code looks good to me, and is very well documented, thank you! 🙌

I've left a few comments, some are minor nits and a couple are more important points to consider.

cdn34dd committed:
Update native android/ios SDK versions to latest version of v2

- iOS 2.30.2
- Android 2.26.2

cdn34dd committed:
Improve handling of dynamic dimensions in SVG elements

- The code now searches within the file itself for the variables used
for width and height and sets those as static values.
- If the variables come from outside, this work is then handled in the
native layer by using the wireframe's data to fill those values.

cdn34dd committed:
Fix typos & rename functions

cdn34dd committed:
Convert SVG constant arrays to Sets & Check SVG element name conversion

cdn34dd committed:
Link jira tickets to potential improvements to the code

cdn34dd committed:
Clean up SVG handling logic in Session Replay and Babel plugin packages for clarity and consistency

OliviaShoup reviewed (dismissed):
looks great. thanks for the update!

cdn34dd committed:
Improve support for SVG properties in style objects & Extract SVG transform properties from style objects

cdn34dd committed:
Add SVG conversion units tests

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+        in context: SessionReplayViewTreeRecordingContext
+    ) -> SessionReplayNodeSemantics? {
+        
+        if (view.accessibilityIdentifier != nil) {
```
If it's the SDK that sets the `accessibilityIdentifier`, we could use a prefix (e.g. `dd-sr-svg-`) so that we can identify which ones to look into, with something like this:
```
guard let identifier = view.accessibilityIdentifier,
          identifier.hasPrefix("dd-sr-svg-") else {
        return nil
    }
```

Otherwise, given the multiple levels of nesting below, I'd recommend to use `guard let` statements instead to exit early and make it more readable. 
 
```suggestion
        guard view.accessibilityIdentifier != nil else {
            return nil
        }
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+            let viewId = context.ids.nodeID(view: view, nodeRecorder: self)
+            let subView = view.subviews[0]
+            
+            if let attrs = view.value(forKey: "attributes") as? [String: String] {
```
```suggestion
            guard let attrs = view.value(forKey: "attributes") as? [String: String] else {
                return nil
            }
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                }
+                
+                let bundle = Bundle(for: SvgViewRecorder.self)
+                if let url = bundle.url(forResource: "assets", withExtension: "bin") {
```
```suggestion
                guard let url = bundle.url(forResource: "assets", withExtension: "bin") {
                    return nil
                }
```

mariedm commented on packages/react-native-session-replay/ios/Sources/DdSessionReplayImplementation.swift:
```diff
@@ -60,9 +65,30 @@ public class DdSessionReplayImplementation: NSObject {
+                let decoder = JSONDecoder()
+                svgMap = try decoder.decode([String: SVGData].self, from: data)
+            } catch {
+                consolePrint("Failed to load or decode assets.json", .debug)
```
Shouldn't we stop here when it fails to load thge assets? We can also add the error to help troubleshooting.

```suggestion
                consolePrint("Failed to load or decode assets.json: \(error)", .debug)
                return
```

mariedm commented on packages/react-native-session-replay/ios/Sources/DdSessionReplayImplementation.swift:
```diff
@@ -60,9 +65,30 @@ public class DdSessionReplayImplementation: NSObject {
         )
-                    
+        
+        let bundle = Bundle(for: DdSessionReplayImplementation.self)
```
Will the assets always be loaded from this bundle?

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+        
+        if (view.accessibilityIdentifier != nil) {
+            let viewId = context.ids.nodeID(view: view, nodeRecorder: self)
+            let subView = view.subviews[0]
```
Can't this fail? Will the view always have at least one subview? In any case, it's safer to safely unwrap it. 

```suggestion
            guard let subView = view.subviews.first else { return }
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                        
+                        var svgAttributes: [String] = []
+                        
+                        if (width == "") {
```
nit 
```suggestion
                        if width.isEmpty {
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
I might have missed them, but do we have correspoding tests for this logic? E.g., testing valid svg data, missing/incorrect data, views that don't have svg data, etc.

cdn34dd replied:
We do have a lot of tests validating the svg data for correctness and to ensure only valid attributes are included, but this is done at the babel plugin level. If we find our custom view with those attributes set at the native level, we know for sure we have a valid SVG to process and that all of the data we expect is there. Otherwise the SVG views are skipped and none of our wrapping takes place.

mariedm commented on packages/react-native-session-replay/ios/Sources/DdSessionReplayImplementation.swift:
```diff
@@ -10,6 +10,11 @@ import DatadogInternal
 import DatadogSDKReactNative
 import React
 
+struct SVGData: Codable {
```
nit: if only used internally + you could also add some doc 
```suggestion
internal struct SVGData: Codable {
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                            imagePrivacyLevel: context.recorder.imagePrivacy
+                        )
+                        
+                        let element = SessionReplaySpecificElement(subtreeStrategy: .ignore, nodes: [
```
nit: could add a comment explaining why we ignore the children

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+            let subView = view.subviews[0]
+            
+            if let attrs = view.value(forKey: "attributes") as? [String: String] {
+                let svgId = context.ids.nodeID(view: subView, nodeRecorder: self)
```
nit: since we rely on multiple hardcoded strings here, we could extract them to make it safer

```
private enum DdSessionReplayConstants {
    static let attributesKey = "attributes"
    static let widthKey = "width"
    static let heightKey = "height"
    static let hashKey = "hash"
    static let typeKey = "type"
    static let svgTypeValue = "svg"
}
```

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                        
+                        guard let svgInfo = svgMap[hash] else {
+                            return nil
+                        }
```
We can move this before the do catch statement, right after getting the hash or after checking the svg type; no need to open the file in this case

mariedm commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                                svg.replaceSubrange(tagEnd, with: dimensions + ">")
+                                svgData = svg
+                            }
+                        }
```
This logic is brittle and prone to break. It could also use some more comments/doc. Could we use a regex instead?

cdn34dd replied:
This is made to handle a very specific scenario, where the user passes dynamic properties to their SVG code in the JS side, but since babel is analyzing these files statically, we can't really resolve the properties if they come from outside the file being parsed. 

So what we do instead is simply remove the properties from the SVG at the babel plugin level and signal that the native layer, which has the actual data, should handle it instead. Native will check the signal, which in this case is if either `width` or `height` is null, and append the values from the wireframes to the end of the SVG tag.

But is this brittle, in theory yes, but in practice is extremely unlikely to break, first because we sanitize all broken/invalid data form the SVG before hand, and the SVG data goes through `SVGO` which ensures it's validity. We are also setting these at the root tag of the svg which does accept that many attributes and finally, if something is broken in the SVG it probably won't render in the user's code anyways.

In any case I'll try it with a regex.

mariedm replied:
It makes sense. I see you added a much more complete PR description since then, so at least now you have documentation on how this all work :)

cdn34dd committed:
Fix linting issues in session replay and babel plugin packages

cdn34dd committed:
Update session replay package android tests

cdn34dd committed:
Update `Podfile.lock` & `yarn.lock` files

cdn34dd committed:
Update Session Replay README.md with SVG support section

marco-saia-datadog commented on packages/react-native-babel-plugin/package.json:
```diff
@@ -39,8 +42,14 @@
+        "@babel/types": "^7.27.1",
+        "fast-glob": "^3.3.3",
+        "svgo": "^4.0.0",
+        "uuid": "^13.0.0"
```
We don't track dependencies brought in transitively via strictly required packages, like those added via `react-native` or `babel`, but for new dependencies that we directly add to our dependencies list, we should include them in https://github.com/DataDog/dd-sdk-reactnative/blob/develop/LICENSE-3rdparty.csv.

marco-saia-datadog commented on packages/react-native-session-replay/ios/Sources/DdSessionReplayImplementation.swift:
```diff
@@ -73,7 +99,7 @@ public class DdSessionReplayImplementation: NSObject {
-
+        
         resolve(nil)
     }
```
```suggestion

        resolve(nil)
    }
```

Sorry 😄 🙏

marco-saia-datadog commented on packages/react-native-session-replay/ios/Sources/SvgViewRecorder.swift:
```diff
@@ -0,0 +1,195 @@
+                if let url = bundle.url(forResource: "assets", withExtension: "bin") {
+                    do {
+                        let fileHandle = try FileHandle(forReadingFrom: url)
+                        defer { try? fileHandle.close() }
```
Nice 💯

marco-saia-datadog commented on packages/react-native-session-replay/src/metro/index.ts:
```diff
@@ -0,0 +1,79 @@
+        try {
+            mergeSvgAssets(assetsDir);
+        } catch (error) {
+            console.warn('[SessionReplayAggregator] merge failed:', error);
```
**ultra-nit**:

```suggestion
            console.warn('[SessionReplayAssetBundler] SVGs merge failed:', error);
```

marco-saia-datadog reviewed (dismissed):
Looks good to me! Thanks for addressing my comments. From my side, the only important change left is to list the 3rd party libraries we have added with these changes. Other than that I think we are good to go 💯

cdn34dd committed:
Improvements to iOS SVG view recorder

cdn34dd committed:
Improve encapsulation and lookup of Session Replay assets on iOS

cdn34dd committed:
Update `LICENSE-3rdparty.csv` dependencies

cdn34dd committed:
Update `build-assets` script in Session Replay package

- Ensure assets directory in Session Replay package is cleaned up when
`yarn prepare` is ran so no unwanted files get packed

cdn34dd committed:
Fix session replay native views not being found by metro

cdn34dd committed:
Improve SvgViewRecorder on iOS

- Only do data checks if the view in question is our custom
SesssionReplayView
- Remove accessibilityIdentifier check as it's not needed
- Update file reading logic to support <iOS 13

cdn34dd committed:
Update default value of SVG extraction in babel plugin to `false`

cdn34dd committed:
Make babel plugin cli generate an 'svg-map.json' file for LocalSVGs

- This will improve performance when the plugin is ran on metro, as it
will first load the data from that file if present, avoiding runtime
parsing

cdn34dd committed:
Update Babel plugin export shape for better CommonJS compatibility

cdn34dd committed:
Add `svg-map` to .gitignore

cdn34dd committed:
Update example apps `Podfile.lock` files

cdn34dd committed:
Downgrade `uuid` version to ensure CommonJS compatibility

cdn34dd committed:
Disable babel plugin `SvgUri` parsing until supported on native SDKs
